frappe/frappe
# perf: role instead of users in workflow actions

Every time any workflow transition is possible we write one workflow action row per user of the role that is permitted to make the transition to the table
When any user performs the action we update his record and delete every other record from the table

A role is what decides if a user should be allowed to make a transition
If we create one row per role we are not losing any information
We are also updating who made the transition in the record

By making this change we will be reducing the number of records that are getting created which are then deleted too. 

- [x] use qb
- [x] ~~write a patch~~ it is not possible to predict what the older state was. Older state is needed to figure out the role to replace the user with. Handling it on the fly 
- [x] ~~Soft delete workflow and cleanup at once so that we do not keep on messing up the indexes all the time~~ Probably not needed anymore
- [x] Check if all the indexes that are set make sense
- [x] Add a child table to store all roles that can perform the transition
- [x] Find better name for Workflow Action Role 

How to test:
1. Create a complicated workflow with multiple roles and transitions
2. Check if the transitions are visible properly in the list view
3. Make transitions by logging in via different roles
4. The states should transition as expected

**Testing backward compatibility**
1. Find a site that has pending workflow transitions
2. Migrate the site to this branch
3. The older transitions should be visible in the workflow transition table 
4. It should be possible to complete those transitions
